lib/install/mutate-into-logical-tree.js

```diff
@@ -70,18 +70,27 @@ function translateTree_ (tree, seen) {
   tree.children.forEach(function (child) {
     pkg.dependencies[moduleName(child)] = translateTree_(child, seen)
   })
-  Object.keys(tree.missingDeps).forEach(function (name) {
+
+  function markMissing (name, requiredBy) {
     if (pkg.dependencies[name]) {
+      if (pkg.dependencies[name].missing) return
       pkg.dependencies[name].invalid = true
```

Comment:
As I say above, these are all `read-installed`'s thing. IMO, the fix for this will be to rewrite `npm ls` to use the `

`invalid` is a dep whose version is incompatible with your package.json.
`extraneous` is a dep that's not in your package.json.
`missing` is a dep in your package.json that's not on disk.

I did a bit of digging for what `realName` is supposed to be. It's non-functional in npm@3 (`read-package-tree` does not currently make this information available), but in npm@2 it was used to report instances where the directory name in `node_modules` didn't match the name in `package.json`. `realName` would be the name of the directory.

In npm@2 this presented this way:
```
invalid@1.0.0 /Users/rebecca/code/npmtest/invalid
└── packagename@1.0.0 (dirname) invalid

npm ERR! invalid: packagename@1.0.0 /Users/rebecca/code/npmtest/invalid/node_modules/dirname
```
In npm@3 that looks like: 
```
invalid@1.0.0 /Users/rebecca/code/npmtest/invalid
├── UNMET DEPENDENCY dirname@1.0.0
└── packagename@1.0.0 extraneous

npm ERR! missing: dirname@1.0.0, required by invalid@1.0.0
npm ERR! extraneous: packagename@1.0.0 /Users/rebecca/code/npmtest/invalid/node_modules/dirname
```